core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -235,4 +415,116 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        GetRequest getRequest = new GetRequest(lookup.getIndex(), lookup.getType(), lookup.getId()).preference("_local").routing(lookup.getRouting());
+        getRequest.copyContextAndHeadersFrom(SearchContext.current());
+        final GetResponse getResponse = client.get(getRequest).actionGet();
+        if (getResponse.isExists()) {
+            List<Object> extractedValues = XContentMapValues.extractRawValues(lookup.getPath(), getResponse.getSourceAsMap());
+            terms.addAll(extractedValues);
+        }
+    }
+
+    static Query handleTermsQuery(List<Object> terms, String fieldName, QueryShardContext context, String minimumShouldMatch, boolean disableCoord) {
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        String indexFieldName;
+        if (fieldType != null) {
+            indexFieldName = fieldType.names().indexName();
+        } else {
+            indexFieldName = fieldName;
+        }
+
+        Query query;
+        if (context.isFilter()) {
+            if (fieldType != null) {
+                query = fieldType.termsQuery(terms, context);
+            } else {
+                BytesRef[] filterValues = new BytesRef[terms.size()];
+                for (int i = 0; i < filterValues.length; i++) {
+                    filterValues[i] = BytesRefs.toBytesRef(terms.get(i));
+                }
+                query = new TermsQuery(indexFieldName, filterValues);
+            }
+        } else {
+            BooleanQuery bq = new BooleanQuery(disableCoord);
+            for (Object term : terms) {
+                if (fieldType != null) {
+                    bq.add(fieldType.termQuery(term, context), BooleanClause.Occur.SHOULD);
+                } else {
+                    bq.add(new TermQuery(new Term(indexFieldName, BytesRefs.toBytesRef(term))), BooleanClause.Occur.SHOULD);
+                }
+            }
+            Queries.applyMinimumShouldMatch(bq, minimumShouldMatch);
+            query = bq;
+        }
+        return query;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (this.fieldName == null) {
+            validationException = addValidationError("field name cannot be null.", validationException);
+        }
+        if (isTermsLookupQuery()) {
+            QueryValidationException exception = termsLookup.validate();
+            if (exception != null) {
+                validationException = QueryValidationException.addValidationErrors(exception.validationErrors(), validationException);
+            }
+        }
+        return validationException;
+    }
+
+    @Override
+    protected TermsQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        TermsQueryBuilder termsQueryBuilder = new TermsQueryBuilder(in.readString());
+        termsQueryBuilder.values = (List<Object>) in.readGenericValue();
```

Comment:
we always read and write the values? Wouldn't it make more sense to do it only in the else of the following if?